includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -1521,51 +1523,127 @@ private function is_amp_allowed_tag( $node ) {
+	 *
+	 * Given a spec name like 'form [method=post]', extract the tag name 'form' and the attributes.
+	 *
+	 * @todo This is admittedly rudimentary. It would be more robust to actually look up the tag spec by the name and obtain the required attributes from there, but this is not necessary yet.
+	 *
+	 * @param string $spec_name Spec name.
+	 * @return array Tag name and attributes.
+	 */
+	private function parse_tag_and_attributes_from_spec_name( $spec_name ) {
+		static $parsed_specs = array();
+		if ( isset( $parsed_specs[ $spec_name ] ) ) {
+			return $parsed_specs[ $spec_name ];
+		}
+
+		$attributes = array();
+
+		/*
+		 * This matches spec names like:
+		 * - body
+		 * - form [method=post]
+		 * - form > div [submit-error][template]
+		 */
+		if ( preg_match( '/^(?P<ancestors>.+? )?(?P<tag_name>[a-z0-9_-]+?)( (?P<raw_attrs>\[.+?\]))?$/i', $spec_name, $matches ) ) {
+			$tag_name = $matches['tag_name'];
+
+			if ( isset( $matches['raw_attrs'] ) ) {
+				$raw_attr_pairs = explode( '][', trim( $matches['raw_attrs'], '[]' ) );
+				foreach ( $raw_attr_pairs as $raw_attr_pair ) {
+					$raw_attr_pair = explode( '=', $raw_attr_pair );
+
+					$attr_key = $raw_attr_pair[0];
+					$attr_val = isset( $raw_attr_pair[1] ) ? $raw_attr_pair[1] : true;
+
+					$attributes[ $attr_key ] = $attr_val;
+				}
+			}
+		} else {
+			$tag_name = strtok( $spec_name, ' ' ); // Fallback case.
+		}
+
+		$parsed_specs[ $spec_name ] = compact( 'tag_name', 'attributes' );
+		return $parsed_specs[ $spec_name ];
+	}
+
 	/**
 	 * Get the first ancestor node matching the specified tag name for the supplied $node.
 	 *
 	 * @since 0.5
+	 * @todo It would be more robust if the the actual tag spec were looked up and then matched for each ancestor, but this is currently overkill.
 	 *
-	 * @param DOMNode $node              Node.
-	 * @param string  $ancestor_tag_name Ancestor tag name.
+	 * @param DOMNode $node               Node.
+	 * @param string  $ancestor_spec_name Ancestor spec name, e.g. 'body' or 'form [method=post]'.
 	 * @return DOMNode|null Returns an ancestor node for the name specified, or null if not found.
 	 */
-	private function get_ancestor_with_tag_name( $node, $ancestor_tag_name ) {
+	private function get_ancestor_with_matching_spec_name( $node, $ancestor_spec_name ) {
+		$parsed_spec_name = $this->parse_tag_and_attributes_from_spec_name( $ancestor_spec_name );
+
 		while ( $node && $node->parentNode ) {
 			$node = $node->parentNode;
-			if ( $node->nodeName === $ancestor_tag_name ) {
+			if ( $node->nodeName === $parsed_spec_name['tag_name'] ) {
+
+				// Ensure attributes match; if not move up to the next node.
+				foreach ( $parsed_spec_name['attributes'] as $attr_name => $attr_value ) {
+					$match = (
+						$node instanceof DOMElement
+						&&
+						true === $attr_value ? $node->hasAttribute( $attr_name ) : strtolower( $node->getAttribute( $attr_name ) ) === $attr_value
+					);
+					if ( ! $match ) {
+						continue 2;
```

Comment:
Good use of `continue 2`, I didn't know about that. It looks like that skips to the next iteration in the `while` loop outside of this `foreach` loop.